lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -361,12 +361,56 @@ struct ExprVisitor {
     for (auto *operand : expr.operands()) {
       auto value = context.convertExpression(*operand);
       if (!value)
-        return {};
+        continue;
+      value = convertToSimpleBitVector(value);
       operands.push_back(value);
     }
     return builder.create<moore::ConcatOp>(loc, operands);
   }
 
+  // Handle replications.
+  Value visit(const slang::ast::ReplicationExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    if (isa<moore::VoidType>(type))
+      return {};
+
+    auto value = context.convertExpression(expr.concat());
+    if (!value)
+      return {};
+    return builder.create<moore::ReplicateOp>(loc, type, value);
+  }
+
+  // Handle single bit selections.
+  Value visit(const slang::ast::ElementSelectExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    auto value = context.convertExpression(expr.value());
+    auto lowBit = context.convertExpression(expr.selector());
+
+    if (!value || !lowBit)
+      return {};
+    return builder.create<moore::ExtractOp>(loc, type, value, lowBit);
+  }
+
+  // Handle range bits selections.
+  Value visit(const slang::ast::RangeSelectExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    auto value = context.convertExpression(expr.value());
+    Value lowBit;
+    if (expr.getSelectionKind() == slang::ast::RangeSelectionKind::Simple) {
+      if (expr.left().constant && expr.right().constant) {
+        auto lhs = expr.left().constant->integer().as<uint64_t>().value();
+        auto rhs = expr.right().constant->integer().as<uint64_t>().value();
+        lowBit = lhs < rhs ? context.convertExpression(expr.left())
+                           : context.convertExpression(expr.right());
+      }
+    } else
+      lowBit = context.convertExpression(expr.left());
+
+    if (!value || !lowBit)
+      return {};
+    return builder.create<moore::ExtractOp>(loc, type, value, lowBit);
```

Comment:
Does this do the right thing for integers or arrays that don't start at `0`? For example in `bit [7:4] x; x[5:4]`, the range select `[5:4]` should select the lowest and second-lowest bit in `x`. Is this shifting part of `ExtractOp`? Or does this have to be handled here?

Reply:
Yes, it can also work. We can select a range of bits arbitrarily only if we follow big endian and little endian. For example, if we select bits like `[4:5]` from`[7:4]`, it will trigger an error: endianness of selection must match declared range (type is 'logic[3:0]')

Reply:
The related IR:
```
%x = moore.variable : !moore.packed<range<bit, 7:4>>
%0 = moore.constant 4 : !moore.int
%1 = moore.extract %x from %0 : !moore.packed<range<bit, 7:4>>, !moore.int -> !moore.packed<range<bit, 5:4>>
```
And when we lower `moore.extract` into `comb.extract`, we will use `comb.shr` to guarantee to extract the correct bits.